### DISCUSSION TITLE
Clock Calibration question

### DISCUSSION THREAD

**Discussion by mechatroniks-git** [SpenceKonde/megaTinyCore#727] [Q&A]
Without any oscillator tuning, I'm running my ATtiny1624 @ 10MHz 3.3V running at room temp. I'm measuring a 1hz pulse with microsecond precision. I did my own calibration on a known clock (1.5 PPM accuracy) and can determine an offset value. 

From device to device, I'm seeing up to ±3500 microseconds offset, within datasheet tolerance of ±4% across the temperature range, but I'm at room temp, it should be ±2%. Once I calibrate it, I store in NV and it's stays fairly close. I need to get within 500 microseconds and I can pretty much do that. 

Every once in a while, I see a device is way off and I have to recalibrate. I'm not sure if this is just drift from time to time or should I expect these to be consistent?

Also, would tuning the oscillator be useful to me? I'm concerned that I'd have to tune at 20MHz @5V and that might not translate to running at 10MHz @ 3.3V.

**Comment by SpenceKonde** [comment]
Remark: You're a LOT better than the +/-2%! 3500 us out of out of 1 million would be +/- well under 1% if ,my math serves me. This s conistent with my observations and the basis of the selfTuning example sketch which:
1. Assumes that at startup the internal HF oscillator can be assumed to be 100% accurate. 
2. Generates an event from the 2^15 kHz internal clock. and times it, thus determining actually frequency of the (highly inaccurate, but stable) 32k internal oscillator. Say we determine that if the main clock is 20 MHz and assume perfect accuracy the 32.768 internal osc is running at 34k. 
3. Using that information, it will then step through the cal options from 0 to 127 (or 63 for the older parts), timing the internal oscillator, looking for the 6 key speeds for a given external oscillator speed, (ex, on 2-series at 20 these are 16 MHz, 20 MHz 24 MHz 25 MHz, 30 MHz and 32 MHz.  (16 MHz 2-series and 20 MHz 1-series drops 32 and adds 12, and 16 1=seroes adds 10. and drops 30. 

That all is done under the assumption that the goal is likely to be overclocking without a crystal (2-series parts esp with extended temp range is likely to work just fine at 32 MHz and room temperature!!  That was the general goal in my mind. It was a surprise to me how much interest there is at making it just run at the rated speed more precisely  That is not what I optimized for and sounds super boring to implement (I am frequently contemplating use cases where I'm tallying up clock cycles and being like "Shit running at 24 MHz this is 20% too slow to meet the constraints of the application. But if I kick it up to 32, I'll have 13% margin.... I guess it's time to make sure I have some E-spec (extended temp range parts which overclock markedly better) parts and grossly violate the manufacturer's specifications.")

Q1: anecdotal reports seem to suggest that the distribution of internal oscillator speeds is not gausian. Most are nearly dead-on-target. A very small number are much further away. There doesn't seem to be anything in the middle. 

Remark: It's a real pity that factory cal stopped calculating and storing the oscillator error term :-(

Q2. I had a very hard time detecting the difference in oscillator speed (with 'scope in measure mode) between 5v and when the chip stopped working entirely. Modern AVRs. in my tests, have virtually no Fosc depencance on voltage ( and I needed a butane torch or freeze spray, not just sticking it in the fridge or a sunbeam to see visible impact from temperature

**Comment by mechatroniks-git** [comment]
You are right, I'm seeing the clocks within 0.35 % off, when I calibrate, not 3.5%.

With my calibration, I can get close to .05 %, which is where I need to be. But some calibrations are better than others, on the same specimen. 

I still need to try and improve on the calibration, which is interrupt driven, to get less variability. I have quite a bit of variability (range of 679 uS over 12 calibrations) of the device's calibration value, such that I'm not sure if it's my calibration code, signal noise or the device's oscillator not being consistent. 

My main question is: Should I expect a device to have a consistent clock every time it boots up? Assuming room temp and voltage in the range of 3-4.2VDC @ 10MHz?

**Reply by SpenceKonde** [reply]
If Tjunction<sub>1</sub> == Tjunction<sub>2</sub>
and Vdd<sub>1</sub> == Vdd<sub>2</sub>, 
and calibration byte<sub>1</sub> == calibration byte<sub>2</sub>

then F<sub>1</sub> == F<sub>2</sub> within a very narrow margin in all my investigations. 

Note that the calibration byte is not stored in NVM unless tuning sketch is run (which stores tuning values in the USERROW, which are loaded if (and only if) a tuned clock speed is selected). Otherwise, writes to calibration register are lost on power cycle. 

I for one have never seen changes large enough to notice for the same specimen under the same conditions even when I'm looking directly at clock speed-related stuff. In fact, I found the change in clock speed over 3.0V < Vdd < 5.0V

Oh, and if this matters in your application, remember that it's the die/junction temperature that you care about. Not the ambient temperature (ex, if you unplug the board, rework something and then plug it back in as soon as the board is cool enough that you can hold without burning yourself, you'll see a different internal oscillator speed than you will once it's cooled to the niormal operating temperature (which is typically just a hair above ambient. If your AVRs are getting hot on their own, you did something wrong, and the chip is now likely damaged. For example, from personal experience (which, on a single lighting project, has cost me nearly a dozen devices), if you have a buck converter that bucks 19V from a laptop power brick down to the 5V needed for MCU + WS2812's and the buck converter's ground becomes disconnected from supply ground while remaining connected to both power rails, while the chip remains connected to the 5V rail and supply ground (in my case, the cause was a defective connector), the full 19v is applied to the chip. On the one occasion that I caught this "in the act, I was holding the PCB, with the chip touching my hand, and suddenly the animation on the lights stopped and the chip became uncomfortably hot to touch in the time it took me to disconnect the board from the rest of the system. That was the night I burned out 6 boards before I gave up on having fancy lighting my roommate's birthday party. While you may question my judgement for continuing to connect fresh PCBs, I was _stone cold sober_ at that party. Really lousy party, too - cops came and the landlord evicted us over it - so losing a half dozen lighting controllers was kind of the high point of the night....)

**Reply by mechatroniks-git** [reply]
I've tightened up my calibration and looks loke what you say is true, its consistant if temp is consistant, which is never a given. 

So I'm going to add in temp compensation. But I'm a couple hundred bytes from full on program flash. How close can I get to 100%?

**Reply by hmeijdam** [reply]
Program flash you can fill up to 100%. Dynamic memory you want to leave sufficient available.

**Comment by mechatroniks-git** [comment]
I noticed that the internal oscillator options for tuned stop at 12MHz, not 10MHz. Is it possible to use internal oscillator at 10MHz tuned? Also, when tuning, can I run it at 3.3V?

**Comment by SpenceKonde** [comment]
the oscillactor usually can't be tuned down that low. though in 16 mhz the first 1-series that I tested could, most of them cant. That option was removed. We now ( as of like 20 mnutes ago on master, but no releases) will tune for 6 of the following: 12, 14, 16, 20, 24, 25, 30, and 32 - if it's 16 MHZ series 1, it's te first 6, if 20 MHz series 2, the last 6, and otherwise the middle 6, But I added prescaled tuning options so you can use tuning when you['re  planning to run at slower speeds can do so - 1, 2, 4, 5, 6, 7 , 8, 10, 12, 14, 16,. 20, 24, 25, 30 (1-series th osccfg @ 20 MHz or 2-series, 32 (2 series only)) (no 7/14 om 2-series, nor supported by the core boards menu at all)

Buit in the process of merging @MX682X's PR and trying to extend it I fucked up bad, and there's no serial. I might be able to look at it weekend, tomorrow not likely. / 

Right now huge changes (different ones entirely) are halfway implemented on both cores and work in neither place in the github branch x_x
